packages/frontend/core/src/modules/workspace/properties/adapter.ts

```diff
@@ -0,0 +1,156 @@
+// the adapter is to bridge the workspace rootdoc & native js bindings
+
+import { createYProxy, type Workspace, type Y } from '@blocksuite/store';
+import { defaultsDeep } from 'lodash-es';
+
+import {
+  PagePropertyType,
+  PageSystemPropertyId,
+  type TagOption,
+  type WorkspaceAffineProperties,
+  type WorkspaceFavoriteItem,
+} from './schema';
+
+const AFFINE_PROPERTIES_ID = 'affine:workspace-properties';
+
+/**
+ * WorkspacePropertiesAdapter is a wrapper for workspace properties.
+ * Users should not directly access the workspace properties via yjs, but use this adapter instead.
+ *
+ * Question for enhancement in the future:
+ * May abstract the adapter for each property type, e.g. PagePropertiesAdapter, SchemaAdapter, etc.
+ * So that the adapter could be more focused and easier to maintain (like assigning default values)
+ * However the properties for an abstraction may not be limited to a single yjs map.
+ */
+export class WorkspacePropertiesAdapter {
+  // provides a easy-to-use interface for workspace properties
+  private readonly proxy: WorkspaceAffineProperties;
+  public readonly properties: Y.Map<any>;
+
+  constructor(private readonly workspace: Workspace) {
+    // check if properties exists, if not, create one
+    const rootDoc = workspace.doc;
+    this.properties = rootDoc.getMap(AFFINE_PROPERTIES_ID);
+    this.proxy = createYProxy(this.properties);
+
+    // fixme: deal with migration issue?
+    this.ensureRootProperties();
+  }
+
+  private ensureRootProperties() {
+    // todo: deal with schema change issue
+    // fixme: may not to be called every time
+    defaultsDeep(this.proxy, {
+      schema: {
+        pageProperties: {
+          custom: {},
+          system: {
+            journal: {
+              id: PageSystemPropertyId.Journal,
+              name: 'Journal',
+              source: 'system',
+              type: PagePropertyType.Date,
+            },
+            tags: {
+              id: PageSystemPropertyId.Tags,
+              name: 'Tags',
+              source: 'system',
+              type: PagePropertyType.Tags,
+              options: this.workspace.meta.properties.tags?.options ?? [], // better use a one time migration
+            },
+          },
+        },
+      },
+      favouriates: {},
```

Comment:
favorite or favourite ?

Reply:
Our App is actually en-GB so favourite > favorite. But it is a matter of the display texts. Not sure if we need to follow the same rules for the internal impl?